packages/components/src/SubHeaderBar/SubHeaderBar.component.js

```diff
@@ -7,85 +8,130 @@ import { DEFAULT_I18N, getDefaultTranslate } from '../translate';
 		<div className={className}>
-			{components.map((component, index) => (
-				<ActionBar.Content
-					key={index}
-					tag={component.tag}
-					left={left}
-					center={center}
-					right={right}
-				>
-					{component.injectedComponent}
-				</ActionBar.Content>
-			))}
+			<ActionBar.Content
+				tag={tag}
+				left={left}
+				center={center}
+				right={right}
+			>
+				{children}
+			</ActionBar.Content>
 		</div>
 	);
 }
 
 SubHeaderBarActions.propTypes = {
-	components: PropTypes.array.isRequired,
+	children: PropTypes.node,
+	tag: PropTypes.string,
 	right: PropTypes.bool,
 	center: PropTypes.bool,
 	left: PropTypes.bool,
-	className: PropTypes.string,
+	hasRight: PropTypes.bool,
 };
 
-function SubHeaderBar({ t, onGoBack, componentsCenter, componentsRight, className, ...rest }) {
-	const defaultComponentsLeft = [
-		{
-			injectedComponent: (
-				<Action
-					id="backArrow"
-					onClick={onGoBack}
-					label={t('BACK_ARROW_TOOLTIP', { defaultValue: 'Go Back' })}
-					icon="talend-arrow-left"
-					bsStyle="link"
-					className={classNames(theme['tc-subheader-back-button'], 'tc-subheader-back-button')}
-					hideLabel
-				/>
-			),
-		},
-		{
-			injectedComponent: <InputTitleSubHeader t={t} {...rest} />,
-		},
-	];
+function CustomInject({ getComponent, left, right, center, nowrap, ...props }) {
+	if (nowrap) {
+		return <Inject getComponent={getComponent} {...props} />;
+	}
+	return (
+		<SubHeaderBarActions left={left} right={right} center={center}>
+			<Inject getComponent={getComponent} {...props} />
+		</SubHeaderBarActions>
+	);
+}
+CustomInject.propTypes = {
+	nowrap: PropTypes.bool,
+	right: PropTypes.bool,
+	center: PropTypes.bool,
+	left: PropTypes.bool,
+	getComponent: PropTypes.func,
+};
+
+function SubHeaderBar({
+	t,
+	onGoBack,
+	components,
+	getComponent,
+	className,
+	children,
+	left,
+	center,
+	right,
+	...rest
+}) {
+	const injected = Inject.all(getComponent, components, CustomInject);
+	const Components = Inject.getAll(getComponent, { Action, ActionBar });
+	const hasRight = (
+		Array.isArray(right) ||
+		has(components, 'before-right') ||
+		has(components, 'after-right')
+	);
 	return (
 		<header className={classNames(theme['tc-subheader'], 'tc-subheader', className)}>
-			<ActionBar className={classNames(theme['tc-subheader-navbar'], 'tc-subheader-navbar')}>
-				<SubHeaderBarActions
-					components={defaultComponentsLeft}
-					className={theme['tc-subheader-navbar-left']}
-					left
-					center={false}
-					right={false}
-				/>
-				{Array.isArray(componentsCenter) && (
-					<SubHeaderBarActions
-						components={componentsCenter}
-						className={classNames([theme['tc-subheader-navbar-center']], 'subheader-center', {
-							[`${theme['tc-subheader-navbar-center-no-margin-right']}`]: componentsRight,
-							'tc-subheader-navbar-center-no-margin-right': componentsRight,
-						})}
-						left={false}
-						center
-						right={false}
-					/>
+			{injected('before-actionbar')}
```

Comment:
it would be even sweeter to have those a react component (moar jsx)

Reply:
I have tried multiple approches and this one seems the better one.
In fact trying to make it as easy as possible (only two lines here).

```javascript
{has(components, 'before-actionbar') && <Inject getComponent={getComponent} {...components['before-actionbar']} />}
```

The other solution to keep it small could be to create an HOC of Inject.
I'm ready to try different approche, please give me a solution !